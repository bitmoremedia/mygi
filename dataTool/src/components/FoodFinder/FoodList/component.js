import React, { PropTypes } from 'react';

const FoodList = ({ list, sourceName, sourceId, associateDataSource, deleteAssociatedDataSource }) => {

  // note: current assumption is that a food can only be associated with one item in the main Food list

  let associatedFoodItem;
  list.forEach((food)=>{
    if (food.sources && food.sources[sourceName] && food.sources[sourceName] === sourceId) {
      associatedFoodItem = food;
    }
  });

  const updateDataSource = (mode, food) => {
    if ( mode === 'add' ){
      associateDataSource({
        "foodId": food.id,
        sourceName,
        sourceId,
      });
    } else if ( mode === 'delete' ){
      deleteAssociatedDataSource({
        "foodId": food.id,
        sourceName,
        sourceId,
      });
    }
  }

  const renderAction = (food) => {
    if (associatedFoodItem) {
      if (food.id === associatedFoodItem.id) {
        return <button onClick={()=>updateDataSource('delete', food)}> - </button>;
      }
      return '';
    }
    return <button onClick={()=>updateDataSource('add', food)}> + </button>;
  };

  return (
    <div>
      <ul>
        {list.map((food)=><li key={food.id}>{food.name} [{food.gi}] {renderAction(food)}</li>)}
      </ul>
    </div>
  );
};

FoodList.propTypes = {
  list: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string.isRequired,
      name: PropTypes.string.isRequired,
      gi: PropTypes.any.isRequired,
    })
  ).isRequired,
};

export default FoodList;
