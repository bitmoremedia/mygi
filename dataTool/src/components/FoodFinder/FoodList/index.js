import React from 'react';

// import { Container } from './styles';

const foodListProps = {
  list: React.PropTypes.arrayOf(
    React.PropTypes.shape({
      id: React.PropTypes.string.isRequired,
      name: React.PropTypes.string.isRequired,
      gi: React.PropTypes.any.isRequired,
    })
  ).isRequired,
};

const FoodList = ({ list }) => {
  return (
    <div>
      <ul>
        {list.map((food)=><li key={food.id}>{food.name} [{food.gi}]</li>)}
      </ul>
    </div>
  );
};

FoodList.propTypes = foodListProps;

export default FoodList;
