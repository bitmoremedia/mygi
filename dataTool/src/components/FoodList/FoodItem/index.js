import React, { PropTypes } from 'react';

import { Item } from './styles';

const FoodItem = ({food, editAction}) => {
  return (
    <Item key={`${food.id}-food-item`}>
      {food.name}
    </Item>
  );
};

FoodItem.propTypes = {
  food: PropTypes.shape({
    id: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired,
    gi: PropTypes.any.isRequired,
  }).isRequired,
};

export default FoodItem;
