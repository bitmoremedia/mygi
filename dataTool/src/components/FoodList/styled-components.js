import styled from 'styled-components'

export const AddButton = styled.div`
  margin-top: 5px;
  margin-bottom: 10px;
`
