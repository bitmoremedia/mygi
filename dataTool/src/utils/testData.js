const testData = {
  foodList: {
    'abcd-123': {
      category: 'Beans',
      gi: 35,
      id: 'abcd-123',
      name: 'Kidney Beans',
      sources: {},
    },
    'efgh-456': {
      category: 'Bread',
      gi: 68,
      id: 'efgh-456',
      name: 'White Bread',
      sources: {
        havard: 'havard-white-bread',
      },
    },
  },
  dataSources: {
    havard: {
      name: 'havard',
      title: 'Havard',
      data: {
        'havard-white-bread': {
          gi: 65,
          id: 'havard-white-bread',
          name: 'White Bread',
        },
      },
    },
    mendosa: {
      name: 'mendosa',
      title: 'Mendosa',
      data: {
        'mendosa-white-bread': {
          gi: 77,
          id: 'mendosa-white-bread',
          name: 'White Bread',
        },
      },
    },
  },
  foodId: 'mendosa-white-bread',
  sourceName: 'mendosa',
}

export default testData
