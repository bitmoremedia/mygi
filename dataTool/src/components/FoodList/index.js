import { connect } from 'react-redux'

import FoodList from './component'

const mapStateToProps = state => ({
  foodList: state.foodList,
  dataSources: state.dataSources,
})

export default connect(mapStateToProps)(FoodList)
