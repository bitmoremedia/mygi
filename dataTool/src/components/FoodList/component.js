import React, { Component } from 'react';
import _values from 'lodash/values';
import _sortBy from 'lodash/sortBy';
import { Link } from 'react-router-dom';

import Table from '../common/Table';
import Modal from '../common/Modal';
import AddEditFood from '../AddEditFood';
import Button from '../common/Button';

import { AddButton } from './styles';
import FoodItem from './FoodItem';

class FoodList extends Component {

  constructor(props, context) {
    super(props, context);
    this.state = {
      showAddFood: false,
      showEditFoodModal: false,
      editFoodId: undefined,
    };
    this.toggleAddFood = this.toggleAddFood.bind(this);
    this.openEditFoodModal = this.openEditFoodModal.bind(this);
    this.closeEditFoodModal = this.closeEditFoodModal.bind(this);
  }

  componentWillReceiveProps(nextProps) {
    // a bit hacky - but if props have changed then hide the add/edit fields
    this.setState({
      showAddFood: false,
      showEditFoodModal: false,
    });
  }

  toggleAddFood() {
    this.setState({
      showAddFood: !this.state.showAddFood
    });
  }

  openEditFoodModal(id) {
    this.setState({
      showEditFoodModal: true,
      editFoodId: id
    });
  }

  closeEditFoodModal() {
    this.setState({
      showEditFoodModal: false,
      editFoodId: undefined
    });
  }

  render() {
    const { toggleAddFood, openEditFoodModal, closeEditFoodModal } = this;
    const { showAddFood, showEditFoodModal, editFoodId } = this.state;
    const { foodList, dataSources } = this.props;
    const foodListArray = _sortBy(_values(foodList), 'name');
    const dataSourcesArray = _sortBy(_values(dataSources), 'title');

    const columns = [{
      key: 'food',
      label: 'Food'
    },{
      key: 'category',
      label: 'Category'
    },{
      key: 'gi',
      label: 'GI Value'
    },{
      key: 'average',
      label: 'Average GI'
    }];
    // add columns for each of our data sources
    dataSourcesArray.forEach((source)=>{
      columns.push({
        key: source.name,
        label: <Link to={`/source/${source.name}`}>{source.title}</Link>,
      });
    });
    const data = foodListArray.map((food)=>{
      const dataColumns = [
        { key: food.id, value: <FoodItem food={food} editAction={openEditFoodModal} /> },
        { key: food.id, value: food.category },
        { key: food.id, value: food.gi },
        { key: food.id, value: 'TBC' },
      ];
      let giTotal = 0, giCount = 0, giAverage;
      // add data columns for each of our data sources
      dataSourcesArray.forEach((source)=>{
        const sourceGi = (food.sources[source.name] && source.data[food.sources[source.name]]) ? source.data[food.sources[source.name]].gi : "";
        dataColumns.push({
          key: `${food.id}-${source.name}`,
          value: sourceGi,
        });
        if (parseInt(sourceGi, 10) > 0) {
          giTotal += parseInt(sourceGi, 10);
          giCount += 1;
        }
      });
      // update the average GI value if we have one
      if ( giTotal > 0 && giCount > 0 ){
         giAverage = (giTotal/giCount);
        dataColumns[3].value = giAverage.toFixed(0);
      }
      return dataColumns;
    });

    const renderToggleAddFoodButton = () => {
      const buttonText = (this.state.showAddFood) ? 'Cancel' : 'Add';
      return <AddButton><Button onClick={()=>toggleAddFood()}>{buttonText}</Button></AddButton>;
    }

    return (
      <div>
        {renderToggleAddFoodButton()}
        {showAddFood && <AddEditFood layout="inline"/>}
        { foodListArray.length > 0 && dataSourcesArray.length > 0 && <Table columns={columns} data={data}/> }
        <Modal visible={showEditFoodModal} onClose={closeEditFoodModal}>
          {showEditFoodModal && <AddEditFood mode='edit' editFoodId={editFoodId} />}
        </Modal>
      </div>
    );
  }
}

export default FoodList;
