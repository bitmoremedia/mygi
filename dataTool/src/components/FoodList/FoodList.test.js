import React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'

import testData from '../../utils/testData'

import FoodList from './component'

test('component renders without error', () => {
  const wrapper = shallow(
    <FoodList
      foodList={testData.foodList}
      dataSources={testData.dataSources}
    />,
  )
  expect(wrapper.exists()).toBe(true)
})

// SNAPSHOT TESTS

test('component renders as expected', () => {
  const wrapper = shallow(
    <FoodList
      foodList={testData.foodList}
      dataSources={testData.dataSources}
    />,
  )
  expect(toJson(wrapper)).toMatchSnapshot()
})
