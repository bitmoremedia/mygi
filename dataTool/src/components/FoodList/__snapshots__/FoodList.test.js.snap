exports[`test component renders as expected 1`] = `
<div>
  <styled.div>
    <Component
      danger={false}
      onClick={[Function]}>
      Add
    </Component>
  </styled.div>
  <Component
    columns={
      Array [
        Object {
          "key": "food",
          "label": "Food",
        },
        Object {
          "key": "category",
          "label": "Category",
        },
        Object {
          "key": "gi",
          "label": "GI Value",
        },
        Object {
          "key": "average",
          "label": "Average GI",
        },
        Object {
          "key": "havard",
          "label": <Link
            replace={false}
            to="/source/havard">
            Havard
      </Link>,
        },
        Object {
          "key": "mendosa",
          "label": <Link
            replace={false}
            to="/source/mendosa">
            Mendosa
      </Link>,
        },
      ]
    }
    data={
      Array [
        Array [
          Object {
            "key": "abcd-123",
            "value": <Component
              editAction={[Function]}
              food={
                      Object {
                              "category": "Beans",
                              "gi": 35,
                              "id": "abcd-123",
                              "name": "Kidney Beans",
                              "sources": Object {},
                            }
              } />,
          },
          Object {
            "key": "abcd-123",
            "value": "Beans",
          },
          Object {
            "key": "abcd-123",
            "value": 35,
          },
          Object {
            "key": "abcd-123",
            "value": "TBC",
          },
          Object {
            "key": "abcd-123-havard",
            "value": "",
          },
          Object {
            "key": "abcd-123-mendosa",
            "value": "",
          },
        ],
        Array [
          Object {
            "key": "efgh-456",
            "value": <Component
              editAction={[Function]}
              food={
                      Object {
                              "category": "Bread",
                              "gi": 68,
                              "id": "efgh-456",
                              "name": "White Bread",
                              "sources": Object {
                                "havard": "havard-white-bread",
                              },
                            }
              } />,
          },
          Object {
            "key": "efgh-456",
            "value": "Bread",
          },
          Object {
            "key": "efgh-456",
            "value": 68,
          },
          Object {
            "key": "efgh-456",
            "value": "65",
          },
          Object {
            "key": "efgh-456-havard",
            "value": 65,
          },
          Object {
            "key": "efgh-456-mendosa",
            "value": "",
          },
        ],
      ]
    } />
  <Modal
    height={280}
    onClose={[Function]}
    visible={false} />
</div>
`;
